src/FSharp.Data.GraphQL.Server.AppInfrastructure/GraphQLWebsocketMiddleware.fs

```diff
@@ -0,0 +1,375 @@
+    }
+
+  let isSocketOpen (theSocket : WebSocket) =
+    not (theSocket.State = WebSocketState.Aborted) &&
+    not (theSocket.State = WebSocketState.Closed) &&
+    not (theSocket.State = WebSocketState.CloseReceived)
+
+  let canCloseSocket (theSocket : WebSocket) =
+    not (theSocket.State = WebSocketState.Aborted) &&
+    not (theSocket.State = WebSocketState.Closed)
+
+  let receiveMessageViaSocket (cancellationToken : CancellationToken) (serializerOptions: JsonSerializerOptions) (executor : Executor<'Root>) (socket : WebSocket) : Task<RopResult<ClientMessage, ClientMessageProtocolFailure> option> =
+    task {
+      let buffer = Array.zeroCreate 4096
+      let completeMessage = new List<byte>()
+      let mutable segmentResponse : WebSocketReceiveResult = null
+      while (not cancellationToken.IsCancellationRequested) &&
+            socket |> isSocketOpen &&
+            ((segmentResponse = null) || (not segmentResponse.EndOfMessage)) do
+        try
+          let! r = socket.ReceiveAsync(new ArraySegment<byte>(buffer), cancellationToken)
+          segmentResponse <- r
+          completeMessage.AddRange(new ArraySegment<byte>(buffer, 0, r.Count))
+        with :? OperationCanceledException ->
+          ()
+
+      let message =
+        completeMessage
+        |> Seq.filter (fun x -> x > 0uy)
+        |> Array.ofSeq
+        |> System.Text.Encoding.UTF8.GetString
+      if String.IsNullOrWhiteSpace message then
+        return None
+      else
+        let! result =
+          message
+          |> deserializeClientMessage serializerOptions
+        return Some result
+    }
+
+  let sendMessageViaSocket (jsonSerializerOptions) (socket : WebSocket) (message : ServerMessage) =
+    task {
+      if not (socket.State = WebSocketState.Open) then
+        if logger.IsEnabled(LogLevel.Trace) then
+          logger.LogTrace("Ignoring message to be sent via socket, since its state is not 'Open', but '{state}'", socket.State)
+      else
+        let! serializedMessage = message |> serializeServerMessage jsonSerializerOptions
+        let segment =
+          new ArraySegment<byte>(
+            System.Text.Encoding.UTF8.GetBytes(serializedMessage)
+          )
+        if not (socket.State = WebSocketState.Open) then
+          if logger.IsEnabled(LogLevel.Trace) then
+            logger.LogTrace("ignoring message to be sent via socket, since its state is not 'Open', but '{state}'", socket.State)
+        else
+          do! socket.SendAsync(segment, WebSocketMessageType.Text, endOfMessage = true, cancellationToken = new CancellationToken())
+
+        if logger.IsEnabled(LogLevel.Trace) then
+          logger.LogTrace("<- Response: {response}", message)
+    }
+
+  let addClientSubscription (id : SubscriptionId) (jsonSerializerOptions) (socket) (howToSendDataOnNext: SubscriptionId -> Output -> Task<unit>) (streamSource: IObservable<Output>) (subscriptions : SubscriptionsDict)  =
```

Comment:
Same here, we could group those function arguments if possible?

Reply:
Fair enough, I grouped some parameters into a tuple. Now the two lines of code that call this function look tidier in my opinion.